### PR TITLE
Added decision cache

### PR DESCRIPTION
**Pull request by vasili-v**

### CODE REVIEW

**Comment by vasili-v** [comment]
The PR contains preview of changes for adding decision cache for now.

**Comment by vasili-v** [comment]
This is now stable version which is ready for review and I don't expect global changes from now.

**Comment by vasili-v** [comment]
One more thing is still missing - options to turn on/off and control cache in Corefile.

File: pdp/errors.yaml [hunk 1]
```diff
@@ -988,3 +991,417 @@ errors:
...
+
+- id: responseEffectError
+  fields:
+  - id: effect
+    type: int
+  msg: "Unknown effect %d"
+  args:
+  - field: effect
+
+- id: ResponseServerError
```
Comment (yuewko) [inline pdp/errors.yaml]: The id `ReponseServerError` should be lowercase.
As it is, the id `ResponseServerErrorID` injected into errors.go becomes exported.

Comment (vasili-v) [reply pdp/errors.yaml]: It's intentional. So code which uses the library can distinct the error by type cast and/or by checking its id.

Comment (yuewko) [reply pdp/errors.yaml]: Oh, yes, of course!

So, we do this on a case by case basis as the need arises? Or is there a convention that I fail to see?

Comment (vasili-v) [reply pdp/errors.yaml]: It's case by case when there is specific need to make this or the need is foreseeable. For the case of `ReponseServerError`, someone may need to know if error returned from PDP or generated locally by for example inability to transfer data.

Additionally, we have a plan to review and make error handling more formal grounded.

Comment (yuewko) [reply pdp/errors.yaml]: Perfect!
Exporting error type on a case by case basis is not ideal, IMHO.
Thanks!

File: pdp/request_test.go [hunk 2]
```diff
@@ -0,0 +1,2093 @@
...
+
+	n, err = MarshalRequestAssignments([]byte{}, testRequestAssignments)
+	assertRequestBufferOverflow(t, "MarshalRequestAssignments", err, n)
+
+	n, err = MarshalRequestAssignments(b[:2], testRequestAssignments)
+	if err == nil {
+		t.Errorf("expected no data put to small buffer but got %d", n)
+	} else if _, ok := err.(*requestBufferOverflowError); !ok {
+		t.Errorf("expected *requestBufferOverflowError but got %T (%s)", err, err)
+	}
```
Comment (yuewko) [inline pdp/request_test.go]: It looks to me `assertRequestBufferOverflow()` used in L42 does exactly what L45-L49 does. Is there any reason why we don't reuse `assertRequestBufferOverflow()` here?

There seems to be other occurrences following too.

Comment (vasili-v) [reply pdp/request_test.go]: Nice catch!

**Pull request merged**